### ISSUE TITLE
Epic: the langres core — one carrier, one op, topology as code

### ISSUE THREAD

**Issue by fxd24** [raisesquad/langres#193]
## Why

**langres should be to entity resolution what PyTorch is to machine learning:** you build the model, and architectures the authors never imagined are expressible without touching the core.

Today it can't be, for one specific reason. `Blocker`, `Comparator`, `Matcher`, `Clusterer` are names *we invented* for positions in a pipeline *we chose* — and they live in the type system. The four components have four different types on the arrows:

```
records → candidates → vectors → judgements → clusters
```

Every arrow lines up exactly one way. That isn't a pipeline we picked; **it's the only pipeline the types permit.**

`nn.Module` composes for exactly one reason: `forward` is `Tensor → Tensor`. Input type equals output type, so any module can follow any module. PyTorch has no `nn.Encoder` base class — a transformer's encoder and decoder are both just `Module`, and `nn.Linear` doesn't know it's layer three. **Position lives in `forward()`, never in the type.** That is precisely why PyTorch could host ResNet, transformers, and diffusion without a core change.

## What we found

**Every defect traces back to one of the four names.** All verified against `origin/main`:

| The name | What it forced | Where |
|---|---|---|
| `Matcher` as a mandatory slot | `EmbeddingScoreMatcher` — a component that computes nothing, copying `similarity_score` → `score` | `core/matchers/embedding_score.py` |
| `Blocker` owning the score | a `type_name == "vector_blocker"` sniff walking the blocker tree on **every** `resolve()` | `core/resolver.py:219` |
| `Clusterer` owning the threshold | `t` passed twice in one topology; closure overrides negatives we already measured | `architectures/vector_llm_cascade.py` |
| Fixed order, each slot once | CLER's matcher-aware blocking depth (PVLDB 17(3)) is **inexpressible** | `core/resolver.py:585` |
| Set-wise work in a pairwise slot | `GroupwiseMatcher` is set-wise **in**, pairwise **out** — the joint decision is flattened away | `core/matcher.py:217` |
| Position in the type | "tune the blocker's embedder" ≠ "tune the matcher's embedder" — two code paths, one operation | — |

**And the four types are already one type.** From `core/models.py`:

```
ERCandidate       = left,    right,     blocker_name,   similarity_score: float|None
PairwiseJudgement = left_id, right_id,  decision_step,  score: float|None, score_type
```

A pair, a score, and a note about who touched it last — twice, in two spellings. `EmbeddingScoreMatcher` is the memcpy between them.

### What the literature says

From a five-agent primary-source review (detail in `docs/THEORY.md`; experiments in the research agenda):

- **The recall-first doctrine is an axiom the field names out loud.** Papadakis et al., ACM CSUR 2020 §2: *"A common assumption in the literature is the **oracle**, i.e., a perfect matching function… This allows for reasoning about the performance of blocking methods **independently of matching methods**."* **Our four names are that axiom compiled into a type system.**
- **"≥95% blocking recall" is folklore** — not found in any primary source. `core/blocker.py:38` asserts it anyway.
- It is **empirically false** where it matters. SC-Block's own Table 5: **8/8 cells on WDC-B where the larger, strictly-higher-recall candidate set produces lower end-to-end F1** — the authors never draw the conclusion. IJCAI 2022 (arXiv 2205.05889) swept the negative ratio 1:3 → 1:100 and Ditto fell **93.06 → 38.52 F1**.
- **Zero papers train a retriever "for recall."** Recall@k is reported, never optimized (ANCE, RocketQA, RocketQAv2, Margin-MSE, TCT-ColBERT, Zhan et al.). What you actually choose is the negative sampling.
- **There is no autograd for ER.** An LLM call, a faiss lookup, and a top-k have no gradients. Credit assignment cannot be automatic — which is why joint training is exactly where the negative results cluster.

## The idea — the lowest common denominator

Three things.

**1. One carrier — `Pairs`.** A scored relation: which pairs are live, what each scores, in what score space. Lazy only at the `A×B` entry (the first π); concrete thereafter, so this needs one index fusion, not a query engine.

**2. One component type — `Op.forward(Pairs) -> Pairs`.** Exactly two operations exist on the carrier:

- **σ** — same rows, new numbers *(rescore)*
- **π** — same numbers, fewer rows *(select)*, at a feasible class **𝓕**

Blocker, matcher, reranker, clusterer become **names for what you're doing**, not types.

**3. The topology is `forward()`** — code, not slots. `dedupe`/`link` are the exit that reads the answer out of the final `Pairs`.

**𝓕 is the parameter that gives one operation seven names:**

| 𝓕 | is called |
|---|---|
| unconstrained | threshold matching |
| `deg_L ≤ k` | top-k blocking |
| `deg_L ≤ 1` | mention→KB linking (+ NIL) |
| `deg ≤ 1` both sides | assignment (Hungarian) |
| equivalence relations | clustering |

**Scope — what an op *mixes over* — is the peer axis.** `nn.Linear` and `nn.MultiheadAttention` take the *same* tensor; one is per-position, one mixes across it. No separate type. Same here:

- `pair` — per-pair (rapidfuzz, a cross-encoder)
- `group` — mixes the rights within a left (a set-wise LLM call)
- `global` — mixes everything (assignment, clustering)

For π, scope is implied by 𝓕. For σ it's a declaration. **Grouping is a view, not new data** — `derive_groups_from_pairs` is already `R.group_by(left)`.

## What it unlocks

None of these are features. They're lines.

```python
# One model does the whole job — no second stage, no null matcher
Sequential(EmbedScore(), Select(equivalence, t=0.9))

# Cascade: an `if`, not a class
def forward(self, R):
    R = self.embed(R)
    return R.update(self.llm(R.between(0.35, 0.85)))

# CLER's matcher-aware depth (PVLDB 17(3)) — a `while`
def forward(self, R):
    while self.deeper(R):                     # the matcher decides how deep the blocker goes
        R = R | self.matcher(self.block(R, k=10))
    return R.select(equivalence, t=self.t)
```

A **reranker** is σ after π. **Collective/iterative ER** is a `while` plus locals — PyTorch never anticipated ResNet either. **Assignment** comes free the day 𝓕 is a parameter, because `scipy.optimize.linear_sum_assignment` already exists. A **set-wise LLM call** is either σ at group scope (score K jointly, all K survive) or π at `𝓕 = deg≤1` (pick one, K−1 rows deleted) — two different ops the current names cannot tell apart.

**langres is the autograd, not the BLAS.** faiss, rapidfuzz, sentence-transformers, litellm, networkx, scipy — each is a σ or π implementation. The core's only job is composition.

## Training — separate, with a global gate

No autograd ⇒ a per-component signal is **forced**, and it is a **surrogate**. LCRON (ICML 2025) gives the rigorous framing: the end-to-end objective is **lower-bounded by the product of per-stage survival probabilities**, so per-stage losses are *auxiliaries that tighten a bound* — not objectives in their own right.

Therefore: **fit locally, gate globally.** Never accept a fit because the local metric improved. Zamani et al. (ICTIR '22) measured an independently-optimized stage 1 scoring **0.6713 vs 0.7160 for not optimizing stage 1 at all**.

- **The tuning unit is the backbone, not the op.** "Fine-tune an embedder" is one operation regardless of where it sits. It attaches to `ModelRef`, which already exists. `encoder → {LoRA, full FT}`, `decoder → {prompt tune, LoRA, full FT}`.
- **A fit's blast radius is everything downstream of it.** Fit `t` → sweep cached scores, free. Fit the matcher → rerun the matcher only. Fit the embedder → *the pairs themselves change*; everything reruns and the index rebuilds.
- **The objective never lives on an op.** Verified precedent: PyTorch's param groups carry only optimizer hyperparameters (`lr`, `momentum`, …) — not one is a loss; the objective enters via `.backward()`, from your loop. DSPy's metric sits on the optimizer, and GEPA's per-component variant keeps **one callable told which component it's scoring**, degrading to program-level. An op declares what it *is*, never what it's *for*.
- **The gate already exists.** The autoresearch `propose → run → eval → keep-if-better` loop, currently scoped to blocking params. Widening its proposal space **is** the training story — no new subsystem.

## Not in scope

- **Joint training / a single joint loss.** No gradient crosses the seam in *any* published ER method — CLER, MutualER and PUER are all alternating co-training over pseudo-labels. And joint training is where the negative results cluster (Sachan's Joint Top-k moved Top-100 *backwards*).
- **Index-refresh infrastructure.** Contested, not settled: four independent groups declined ANCE-style refresh and matched or beat it. ADORE trains only the query side against a **frozen** index; the query-side gain is real, though the 0.347 vs ANCE's 0.338 MRR@10 is the ADORE**+STAR** combination (bare ADORE trails ANCE on several rows) — verify per-row before quoting.
- **Promising an argmax at 𝓕 = equivalence.** Weighted correlation clustering has no believed constant-factor approximation (Dedupalog's MULTICUT argument). It ships labeled *heuristic*, with the algorithm named.
- **Assuming a shared blocker/matcher encoder works.** It has a structural obstacle — blocking needs decomposable, precomputable per-record vectors; matching deliberately uses cross-attention over the pair — and zero prior art. Sudowoodo uses *different base models* for the two and argues against sharing. Open research, not a design assumption.

## Open questions

- **Does the merge/fuse exit forfeit Swoosh's ICAR?** Representativity licenses merge-on-match, but *"confidences may make order critical"* — our merge carries scores, so `dedupe()` may be order-dependent. Decide consciously rather than discover it.
- **Multi-relation carrier for collective ER.** A single `A×B` cannot express cross-type constraints (Dedupalog). Cheap to allow now, expensive after every signature is written.
- **The score-space seam.** A cascade's rows carry *different* score types — `VectorLLMCascade` already refuses calibration for exactly this reason, in prose. Union across spaces should be a wiring-time error, not a docstring.

## Sequence

1. **Unify the carrier** (`ERCandidate` + `PairwiseJudgement` → `Pairs`) — mostly mechanical; the structs are already identical.
2. **`Op` replaces the role types** — `Blocker`/`Matcher`/`Clusterer` become names, not types.
3. **`forward()` replaces the four slots** — the only breaking move.
4. **Backbone as the tuning unit.**
5. **Widen keep-if-better** from blocking params to any fit.

1→2→3 is the inference half and strictly ordered — the carrier unblocks the rest. 4→5 is training, and 5 already exists, scoped.

---

*Design record: `docs/THEORY.md` (the algebra, §12 errata). Experiments this unlocks: the research agenda doc. Evidence base: five parallel primary-source literature reviews, 2026-07-17.*

**Comment by fxd24** [comment]
## Progress — the "proven core" is built and up for landing (PR #217)

The load-bearing, breaking part of this epic — the carrier + the op algebra + the generalized spine + persistence — is **complete, verified, and open as [PR #217](https://github.com/fxd24/langres/pull/217)** (`epic/193-core-algebra` → `main`). It ships the seam and *proves* it with a previously-inexpressible architecture, while keeping every existing model byte-identical.

### What landed on the integration branch (17 commits, PRs #203–#216)

**The algebra is real, not just in `THEORY.md`:**
- **`core/pairs.py`** — the one carrier `Pairs`: id-referenced rows bound to a record store, lazy `.left`/`.right`, one unified `score`/`score_type` column (`score_type=None` = "blocked, not yet scored").
- **`core/op.py`** — the one component type `Op.forward(Pairs) -> Pairs`: `Score` (rescore) and `Select` (select at a feasible class) + concrete `ThresholdSelect`/`TopKSelect`. `Sequential.check()` enforces the theory's guardrails at **wiring time** (a select over a non-orderable vector score is a build-time error).
- **`core/op_adapters.py`** — the seven role adapters (`BlockerSource`, `ComparatorScore`, `MatcherScore`, `GroupwiseMatcherScore`, `ClustererStage`, `CanonicalizeFinalize`), so every existing component runs as an `Op`.

**The spine composes topology as code:**
- A public **`from_topology(ops=...)`** door + an optional explicit `_ops` chain the spine walks; the classic four-slot path is kept and dispatched unchanged.
- **Persist v2** — explicit chains round-trip (`artifact_version="2"`); classic stays `"1"` and byte-identical.
- The spend cap is **enforced structurally** on any explicit chain (every billing `Score` is auto-capped through the model's one monitor; a declared `Spending` allowlist replaced the old rot-open denylist, and it now reads both `cost_usd` and `llm_cost_usd`).

### The payoff, proven end-to-end

**`architectures/reranker.py`** — a **Score-after-Select** topology (cheap first pass → top-k prune → rich rescore of the survivors). That shape is *inexpressible* in four fixed slots; it's built here entirely through the public `from_topology` door with **zero core change**, runs offline at $0, and save/loads. A non-vacuous golden proves the second score changes the outcome (a pass-1-only baseline over-merges a same-name/different-address pair; the rerank demotes it). `VectorLLMCascade.save()` also round-trips now (was `NotImplementedError`).

### Why nothing existing breaks

Widen-then-narrow: the four slots were **kept** as the source of truth, so every classic model is byte-identical — `recipe_id` stable, `artifact_version="1"`, old v1 artifacts load unchanged. Guaranteed by a frozen **behavior-parity net** (goldens captured from the pre-refactor tree, re-asserted after every wave) + legacy-v1 load fixtures.

### Verification (beyond green CI)

- Integrated suite on the merged tip: **3849 pass / 98.27%**, parity net **46 pass**, import discipline **87 pass**, **mypy --strict + ruff clean**, and a **two-process fresh-interpreter** Reranker save→load→re-dedupe → identical clusters.
- **Cross-model review (codex, independent):** PR-A took **3 rounds** — each caught a money-guarantee gap that green CI + the goldens + solo review all missed; PR-B clean; PR-C confirmed *genuinely proven and additive* (zero core diff).

### Wave status

| Wave | State |
|---|---|
| W1 — unify the carrier (`Pairs`) | ✅ done |
| W2 — `Op` replaces the role types | ✅ done |
| W3 — `forward()` / topology-as-code | ✅ goal achieved (additive `from_topology`); deleting the four slots is deliberately deferred to W5 |
| W4 — backbone as the tuning unit | ⬜ next |
| W5 — widen keep-if-better (wire `fit` into the optimizer loop) + delete slots | ⬜ next |

The hard two-thirds — the risky breaking rewrite — is done. **W4 + W5 (connect `fit`/the optimizer loop to the new seam) + the cosmetic tail (slot deletion, doc reconciliation, the `_ops`-model `.schema` nit) become a focused follow-up epic** off freshly-landed `main`.

_Once PR #217 is confirmed green it's ready to merge._

**Comment by fxd24** [comment]
## Update — research execution foundation is ready in PR #229

PR [#229](https://github.com/fxd24/langres/pull/229) builds the research, benchmarking, measurement, reproduction, tracking, and Hugging Face lifecycle on top of this epic’s proven inference core.

### What is now available

- role-neutral `Embedder`, `Reranker`, and `LLM` resources
- composable `Retrieve`, `Rerank`, `Generate`, `Parse`, `Select`, and downstream `Cluster` operations
- four standard recipes: `Retrieve`, `RetrieveRerank`, `RetrieveLLM`, and `RetrieveRerankLLM`
- reproducible experiment matrices with train/test threshold discipline, cache/resume, failure isolation, and optional shared budgets
- standardized quality, funnel, token, cache/reasoning-token, cost, latency, runtime, hardware, and model-provenance measurements
- local run persistence plus optional Trackio publication
- safe `save_pretrained`, `from_pretrained`, and `push_to_hub` architecture sharing
- zero-network examples, reproduction bundles, repricing, and generated documentation tables

### Epic status

This does **not** close #193.

- W1 — unified `Pairs` carrier: done
- W2 — common `Op` algebra: done
- W3 — topology as code: done and now exercised by the research layer
- W4 — backbone / `ModelRef` as the complete tuning unit: still outstanding
- W5 — widen `propose → run → evaluate → keep-if-better` beyond blocking parameters: still outstanding

The legacy four-slot architecture also remains as an intentional compatibility layer; this work did not delete it. The remaining score-space, multi-relation-carrier, and merge-order questions are likewise still open.

In short: the composable inference core is built and now has a standardized research execution system around it. The generalized training and optimization half remains the next part of this epic.

**Comment by fxd24** [comment]
## Follow-up: embedding resource coverage

The resource seam should ultimately make these embedding variants equally easy to select, benchmark, persist, and share:

- Native Sentence Transformers models.
- Any Hugging Face encoder with an explicit mean/CLS pooling strategy.
- Fine-tuned local embedding checkpoints.
- Quantized or ONNX/OpenVINO variants.

PR #229 provides the role-neutral `Embedder` contract and a first-class `SentenceTransformer` adapter, so researchers can already inject a custom implementation. The remaining DX gap is first-class adapters/configuration for the other variants above, with the same `ModelRef`, runtime facts, measurement, experiment, cache, and Hub lifecycle contracts. This should be treated as resource coverage rather than new retrieval architectures: swapping the embedding implementation must not change the recipe topology.
